sound/soc/sof/pcm.c

```diff
@@ -110,10 +110,14 @@ static int sof_pcm_dsp_pcm_free(struct snd_pcm_substream *substream,
 	/* send IPC to the DSP */
 	ret = sof_ipc_tx_message(sdev->ipc, stream.hdr.cmd, &stream,
 				 sizeof(stream), &reply, sizeof(reply));
-	if (!ret)
-		spcm->prepared[substream->stream] = false;
+	if (ret)
+		return ret;
 
-	return ret;
+	spcm->prepared[substream->stream] = false;
+
+	/* free pipeline components */
+	return sof_pipeline_free(sdev, spcm->stream[substream->stream].list, spcm,
+				substream->stream);
```

Comment:
As @plbossart is advocating (and I agree with that) in forked pipeline cases we should have branch_number + 1 pipelines where the common part (the fork handle :-) ) is one pipeline and all fork teeth are other pipelines. But ATM we have pipelines where one of the teeth has the same pipeline ID as the handle. Wouldn't this free such a pipeline leaving all other teeth orphan? Or do we plan to fix all those pipelines?

Reply:
@lyakh maybe the naming is confusing. pipeline_free doesnt really mean free the pipeline from host->dai. It means free all widgets and pipelines that are needed for that PCM as shown by snd_soc_dapm_dai_get_connected_widgets(). I can update the naming in the next version.